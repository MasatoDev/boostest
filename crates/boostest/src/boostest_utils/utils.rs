use anyhow::{anyhow, Result};
use std::path::{Path, PathBuf};

use glob::glob;
use std::collections::HashMap;
use std::ffi::OsStr;
use std::fs::{self, File};
use std::io::{self, prelude::*};

pub fn read(path: &Path) -> io::Result<String> {
    let mut f = File::open(path)?;
    let mut s = String::new();
    match f.read_to_string(&mut s) {
        Ok(_) => Ok(s),
        Err(e) => Err(e),
    }
}

pub fn read_matching_files(
    patterns: Vec<String>,
    out_file_name: &String,
) -> anyhow::Result<HashMap<PathBuf, String>> {
    let mut contents: HashMap<PathBuf, String> = HashMap::new();

    for pattern in &patterns {
        let result_glob = glob(pattern)?;

        for entry in result_glob {
            let path = entry?;

            if path
                .file_name()
                .unwrap_or(OsStr::new("none_file_name"))
                .to_string_lossy()
                .contains(out_file_name.as_str())
            {
                continue;
            }

            if path.is_file() {
                contents.insert(path.clone(), fs::read_to_string(path)?);
            }
        }
    }

    Ok(contents)
}

pub fn find_boostest_json_recursive(dir: PathBuf) -> anyhow::Result<PathBuf> {
    for entry in fs::read_dir(dir)? {
        let entry = entry?;
        let path = entry.path();

        if path.is_file() {
            if let Some(file_name) = path.file_name() {
                if file_name == "boostest.setting.json" {
                    return Ok(path);
                }
            }
        } else if path.is_dir() {
            if let Ok(found_path) = find_boostest_json_recursive(path) {
                return Ok(found_path);
            }
        }
    }

    Err(anyhow!("boostest.json not found"))
}

/**
 * Normalize and resolve the path.
 * parent: "hoge/a/b/c", relative: "../../d/e/f" => "hoge/a/d/e/f"
 */
pub fn normalize_and_resolve_path(parent: &Path, relative: &Path) -> Result<PathBuf, String> {
    let combined_path = parent.join(relative);
    let resolved_path = combined_path
        .components()
        .fold(PathBuf::new(), |mut acc, comp| {
            match comp {
                std::path::Component::ParentDir => {
                    acc.pop();
                }
                std::path::Component::CurDir => {}
                other => acc.push(other),
            }
            acc
        });

    if resolved_path.is_absolute() {
        Ok(resolved_path)
    } else {
        Err("The resolved path is not absolute.".to_string())
    }
}
