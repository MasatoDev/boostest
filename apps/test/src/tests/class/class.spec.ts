import { runSnapshotTest } from "../utils";
import { ArgObjClass } from "../../types/pattern/arg_obj_class";
import { ArgObjClass2 } from "../../types/pattern/arg_obj_class2";
import {
  ConstructorSignature,
  InnerConstructorSignature,
  ConstructorSignatureInterface,
} from "../../types/ts_types/constructor_signature";
import { boostestArgObjClass } from "./boostest_output/boostestArgObjClass";
import { boostestArgObjClass2 } from "./boostest_output/boostestArgObjClass2";
import { boostestConstructorSignature } from "./boostest_output/boostestConstructorSignature";
import { boostestInnerConstructorSignature } from "./boostest_output/boostestInnerConstructorSignature";
import { boostestConstructorSignatureInterface } from "./boostest_output/boostestConstructorSignatureInterface";
import { boostestCallSignatureInterface } from "../call_signature/boostest_output/boostestCallSignatureInterface";
import { CallSignatureInterface } from "@/ts_types/call_signature";

describe("Class Tests", () => {
  runSnapshotTest("ArgObjClass", boostestArgObjClass<ArgObjClass>());
  runSnapshotTest("ArgObjClass2", boostestArgObjClass2<ArgObjClass2>());

  // FIXME:CallSignatureInterface
  runSnapshotTest(
    "CallSignatureInterface",
    boostestCallSignatureInterface<CallSignatureInterface>(),
  );
  runSnapshotTest(
    "ConstructorSignature",
    boostestConstructorSignature<ConstructorSignature>(),
  );
  runSnapshotTest(
    "InnerConstructorSignature",
    boostestInnerConstructorSignature<InnerConstructorSignature>(),
  );
  runSnapshotTest(
    "ConstructorSignatureInterface",
    boostestConstructorSignatureInterface<ConstructorSignatureInterface>(),
  );
});
