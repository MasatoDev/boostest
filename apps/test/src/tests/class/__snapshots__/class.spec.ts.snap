// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Class Tests ArgObjClass matches snapshot 1`] = `"{"name":"test string data","age":10,"obj_key":{"obj_name":"test string data","obj_val":10},"obj_key2":{"obj_name":"test string data","obj_val":10}}"`;

exports[`Class Tests ArgObjClass2 matches snapshot 1`] = `"{"name":"test string data","age":10,"obj_key":{"obj_name":"test string data","obj_val":10},"obj_key2":{"obj_name":"test string data","obj_val":10}}"`;

exports[`Class Tests CallSignatureInterface matches snapshot 1`] = `"(name, age) => { }"`;

exports[`Class Tests ConstructorSignature matches snapshot 1`] = `
"class ref_918c86bdc296277db65e6efd48c98da06b355275e634c30e981be851125588ff {
    name;
    constructor(name) {
        this.name = name;
    }
}"
`;

exports[`Class Tests ConstructorSignatureInterface matches snapshot 1`] = `
"class ref_918c86bdc296277db65e6efd48c98da06b355275e634c30e981be851125588ff {
    name;
    constructor(name) {
        this.name = name;
    }
}"
`;

exports[`Class Tests InnerConstructorSignature matches snapshot 1`] = `
"class ref_918c86bdc296277db65e6efd48c98da06b355275e634c30e981be851125588ff {
    name;
    constructor(name) {
        this.name = name;
    }
}"
`;
